Update material dialog library

ShridharGoel opened the pull request.

david-allison reviewed (changes requested):
```
> Task :AnkiDroid:deleteSingleFileInGradle UP-TO-DATE
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.kt: (24, 39): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.kt: (82, 27): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.kt: (89, 27): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.kt: (145, 27): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.kt: (253, 27): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.kt: (276, 27): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.kt: (346, 27): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.kt: (400, 27): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.kt: (411, 27): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.kt: (473, 27): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.kt: (499, 27): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/CardTemplateEditorTest.kt: (604, 27): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.kt: (31, 39): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.kt: (220, 45): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.kt: (222, 16): Unresolved reference: getActionButton
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.kt: (235, 38): Unresolved reference: contentView
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.kt: (242, 23): Unresolved reference: contentView
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CreateDeckDialogTest.kt: (22, 39): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CreateDeckDialogTest.kt: (136, 47): Unresolved reference: getActionButton
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CreateDeckDialogTest.kt: (136, 63): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CreateDeckDialogTest.kt: (138, 77): Unresolved reference: inputEditText
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt: (21, 39): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt: (71, 22): Unresolved reference: getActionButton
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt: (71, 38): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt: ([121](https://github.com/ankidroid/Anki-Android/runs/7054575074?check_suite_focus=true#step:8:122), 47): Unresolved reference: items
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt: (121, 54): Type mismatch: inferred type is Matcher<(Mutable)Iterable<*>!>! but Boolean was expected
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/DeckPickerContextMenuAndroidTest.kt: (38, 35): Unresolved reference: items
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (28, 39): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (62, 33): Unresolved reference: customView
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (67, 20): Unresolved reference: getActionButton
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (67, 36): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (93, 33): Unresolved reference: customView
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (98, 20): Unresolved reference: getActionButton
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (98, 36): Unresolved reference: DialogAction
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (121, 33): Unresolved reference: customView
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (156, 33): Unresolved reference: customView
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (194, 33): Unresolved reference: customView
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (243, 33): Unresolved reference: customView
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (290, 33): Unresolved reference: customView
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (344, 33): Unresolved reference: customView
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (388, 33): Unresolved reference: customView
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (429, 33): Unresolved reference: customView
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/tags/TagsDialogTest.kt: (473, 33): Unresolved reference: customView
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/dialogs/utils/RecursivePictureMenuUtil.kt: (30, 27): Unresolved reference: recyclerView
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/ui/KeyPickerTest.kt: (21, 31): Unresolved reference: KeySelectionDialogBuilder
e: /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/ui/KeyPickerTest.kt: (68, 41): Unresolved reference: KeySelectionDialogBuilder
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.kt:
```diff
@@ -15,9 +15,12 @@
  * You should have received a copy of the GNU General Public License along with         *
  * this program.  If not, see <http://www.gnu.org/licenses/>.                           *
  ****************************************************************************************/
+@file:Suppress("Deprecation") // ProgressDialog deprecation
```
can we suppress closer to the actual instantiation (via removing the import and fully qualifying it where it's used)

mikehardy commented:
Getting this in would be incredible - this is on my short list of things I'd love to see merge ASAP (if possible of course)

lukstbit commented:
Reviewed some of the changes and found:
- `iconAttr(R.attr.dialogErrorIcon)` can't be replaced by `icon(R.attr.dialogErrorIcon)` because it would crash as the new library code expects to find either a `DrawableRes` or a `Drawable`. We need to get the drawable from the theme with something like :
```
val ta = contextUsedForDialog.obtainStyledAttributes(
    IntArray(1).apply { this[0] = R.attr.dialogErrorIcon }
)
val dialogrIcon = ta.getDrawable(0) // I think we could get the drawble identifier with getResourceId(0)
ta.recycle()
icon(drawable = dialogErrorIcon)
```
Also the update made this icon bigger for some reason.
- the update breaks the button color for the black theme, dialog button text is colored with black on a black background(Go to NoteEditor -> change a bit the field text -> hit back to trigger dialog-> see difference between before/after library update).

BrayanDSO commented:
> Reviewed some of the changes and found:
> 
> * `iconAttr(R.attr.dialogErrorIcon)` can't be replaced by `icon(R.attr.dialogErrorIcon)` because it would crash as the new library code expects to find either a `DrawableRes` or a `Drawable`. We need to get the drawable from the theme with something like :
> 
> ```
> val ta = contextUsedForDialog.obtainStyledAttributes(
>     IntArray(1).apply { this[0] = R.attr.dialogErrorIcon } // the attr used in AbstractFlashcardViewer
> )
> val dialogrIcon = ta.getDrawable(0) // I think we could get the drawable identifier with getResourceId(0)
> ta.recycle()
> icon(drawable = dialogErrorIcon)
> ```
> 
> Also the update made this icon bigger for some reason.
> 
> * the update breaks the button color for the black theme, dialog button text is colored with black on a black background(Go to NoteEditor -> change a bit the field text -> hit back to trigger dialog-> see difference between before/after library update).

Haven't tested, but I suspect that https://github.com/ankidroid/Anki-Android/blob/main/AnkiDroid/src/main/java/com/ichi2/themes/Themes.kt#L145-L148 may be useful here

david-allison commented:
Just want to say that I love this and want to get it in ASAP to avoid conflicts

lukstbit reviewed (changes requested):
Reviewed the files, mostly ok with some issues:
- why replace in some cases MaterialDialog with the deprecated platform ProgressDialog?
- some unused constants added to ModelEditorContextMenu?
- why the need for extra context parameter in ReadText?
- GestureSelectionDialogUtils is not implemented correctly because its second method calls the first one which always creates a new instance of GesturePicker. So, in the code at first we call getGesturePicker() to get a new instance to use but when calling onGestureChanged() we set a listener on a new instance of GesturePicker and not on the one we previously created. Should probably be handled how KeySelectionDialogBuilder was handled.
- the RobolectrictTest.getDialogtest() is not implemented correctly, should be something like this:
` dialog.view.contentLayout.findViewById<TextView>(R.id.md_text_message).text.toString()`

ShridharGoel commented:
> why replace in some cases MaterialDialog with the deprecated platform ProgressDialog?

New version doesn't have progress support.

> some unused constants added to ModelEditorContextMenu?

They were part of older code, removed now.

> why the need for extra context parameter in ReadText?

It's required for `getResFromAttr` method.

> GestureSelectionDialogUtils is not implemented correctly because its second method calls the first one which always creates a new instance of GesturePicker. So, in the code at first we call getGesturePicker() to get a new instance to use but when calling onGestureChanged() we set a listener on a new instance of GesturePicker and not on the one we previously created. Should probably be handled how KeySelectionDialogBuilder was handled.

Fixed.

> the RobolectrictTest.getDialogtest() is not implemented correctly, should be something like this:
dialog.view.contentLayout.findViewById<TextView>(R.id.md_text_message).text.toString()

Fixed, thanks.

david-allison reviewed (changes requested):
```
> Task :AnkiDroid:compilePlayDebugAndroidTestJavaWithJavac FAILED
/Users/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/androidTest/java/com/ichi2/anki/DeckPickerTest.java:107: error: cannot find symbol
        onView(withId(R.id.md_buttonDefaultPositive)).perform(click());
                          ^
  symbol:   variable md_buttonDefaultPositive
  location: class id
1 error
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CardBrowserMySearchesDialog.kt:
```diff
@@ -56,49 +58,46 @@ class CardBrowserMySearchesDialog : AnalyticsDialogFragment() {
-                .positiveText(getString(android.R.string.ok))
-                .negativeText(getString(R.string.dialog_cancel))
-                .input(R.string.card_browser_list_my_searches_new_name, R.string.empty_string) { _: MaterialDialog?, text: CharSequence ->
+            dialog.title(text = getString(R.string.card_browser_list_my_searches_save))
+                .positiveButton(android.R.string.ok)
+                .negativeButton(R.string.dialog_cancel)
+                .input(hintRes = R.string.card_browser_list_my_searches_new_name) { _: MaterialDialog?, text: CharSequence ->
                     Timber.d("Saving search with title/terms: %s/%s", text, currentSearchTerms)
                     mySearchesDialogListener!!.onSaveSearch(text.toString(), currentSearchTerms)
                 }
         }
```
This is missing `R.string.empty_string`. Is it necessary?

ShridharGoel replied:
It's not required here. Empty string is the default value.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CardBrowserMySearchesDialog.kt:
```diff
@@ -56,49 +58,46 @@ class CardBrowserMySearchesDialog : AnalyticsDialogFragment() {
                     notifyAdapterDataSetChanged()
                 }
-                dialog.dismiss()
+                dialog?.dismiss()
```
May be useful to comment here or extract to a method: dismiss the root dialog

ShridharGoel replied:
What do you mean by extracting to method here?

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CardBrowserOrderDialog.kt:
```diff
@@ -39,27 +40,33 @@ class CardBrowserOrderDialog : AnalyticsDialogFragment() {
+        return MaterialDialog(requireActivity()).show {
+            title(R.string.card_browser_change_display_order_title)
+            message(R.string.card_browser_change_display_order_reverse)
+            listItemsSingleChoice(
+                items = items.toList(),
+                initialSelection = requireArguments().getInt("order")
+            ) {
+                    materialDialog: MaterialDialog, index: Int, charSequence: CharSequence ->
+                orderSingleChoiceDialogListener?.invoke(materialDialog, index, charSequence)
+            }
```
```suggestion
            listItemsSingleChoice(
                items = items.toList(),
                initialSelection = requireArguments().getInt("order"),
                selection = orderSingleChoiceDialogListener
            )
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CardBrowserOrderDialog.kt:
```diff
@@ -39,27 +40,33 @@ class CardBrowserOrderDialog : AnalyticsDialogFragment() {
 
     companion object {
-        private var mOrderDialogListener: ListCallbackSingleChoice? = null
+        private var orderSingleChoiceDialogListener: SingleChoiceListener = null
```
`KotlinCleanup`: lateInit

ShridharGoel replied:
Not possible since `SingleChoiceListener` is nullable.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/CreateDeckDialog.kt:
```diff
@@ -52,23 +53,24 @@ class CreateDeckDialog(private val context: Context, private val title: Int, pri
-        displayKeyboard(show.inputEditText!!, show)
-        mShownDialog = show
-        return show
+        val dialog = MaterialDialog(context).show {
+            title(title)
+            positiveButton(R.string.dialog_ok) {
+                onPositiveButtonClicked()
+            }
+            negativeButton(R.string.dialog_cancel)
+            input(prefill = mInitialDeckName)
+            displayKeyboard(getInputField())
+        }
+        mShownDialog = dialog
```
This seems to be missing `inputRange`

ShridharGoel replied:
`inputRange(1, -1)` is the default, so not required here. Also, updated library doesn't have input range. It only has maximum value (which has the confusing name of `maxLength`) as far as I know.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:
```diff
@@ -61,41 +66,41 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
+                    neutralButton(R.string.close) {
+                        exit()
+                    }
+                    getCustomView().findViewById<View>(R.id.md_button_negative).isEnabled =
+                        (activity as DeckPicker?)!!.hasErrorFiles()
```
Needs testing:

```
                    setActionButtonEnabled(WhichButton.POSITIVE, (activity as DeckPicker?)!!.hasErrorFiles())
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/DatabaseErrorDialog.kt:
```diff
@@ -322,23 +322,25 @@ class DatabaseErrorDialog : AsyncDialogFragment() {
                         }
-                    }
-                    .show()
+                    // .itemsColor(ContextCompat.getColor(requireContext(), R.color.material_grey_500))
```
Let's remove this

david-allison reviewed (commented):
I got up to [DeckPickerBackupNoSpaceLeftDialog.kt](https://github.com/ankidroid/Anki-Android/pull/11646/files?w=1#diff-76e25697508bdbd8d24cf1ce66861ab2620a7b1e63e8f4671a80836f189f4988)

Thanks so much for this. I'm well aware of the effort, and want this in as soon as possible to reduce the chance of conflicts

david-allison commented on AnkiDroid/src/test/java/com/ichi2/anki/dialogs/CustomStudyDialogTest.kt:
```diff
@@ -118,7 +120,8 @@ class CustomStudyDialogTest : RobolectricTest() {
         scenario.onFragment { f: CustomStudyDialog ->
             val dialog = f.dialog as MaterialDialog?
             MatcherAssert.assertThat(dialog, IsNull.notNullValue())
-            MatcherAssert.assertThat(dialog!!.items, Matchers.not(Matchers.hasItem(getResourceString(R.string.custom_study_increase_new_limit))))
+            // TODO: Update this for new material dialog
+            // MatcherAssert.assertThat(dialog!!.items, Matchers.not(Matchers.hasItem(getResourceString(R.string.custom_study_increase_new_limit))))
```
This was /horrible/ to work through. Not used to reflection in Kotlin

```kotlin
/*
 *  Copyright (c) 2022 David Allison <davidallisongithub@gmail.com>
 *
 *  This program is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free Software
 *  Foundation; either version 3 of the License, or (at your option) any later
 *  version.
 *
 *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along with
 *  this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.ichi2.testutils

import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.list.getListAdapter
import kotlin.reflect.full.declaredMemberProperties
import kotlin.test.assertNotNull

/**
 * Extension method to obtain the items from a `MaterialDialog`
 */
val MaterialDialog.items: ArrayList<CharSequence>  
    get() {
        @Suppress("UNCHECKED_CAST")
        return this.getListAdapter()?.let { adapter ->
            val itemsMethod = adapter::class.declaredMemberProperties.find { x -> x.name == "items" }
            assertNotNull(itemsMethod)
            itemsMethod.call(adapter)
        } as ArrayList<CharSequence>
    }
```

david-allison commented on AnkiDroid/src/test/java/com/ichi2/anki/dialogs/DeckPickerContextMenuAndroidTest.kt:
```diff
@@ -34,14 +31,15 @@ class DeckPickerContextMenuAndroidTest : RobolectricTest() {
         // "Delete deck" was previously close to "Custom study" which caused misclicks.
         // This is less likely at the bottom of the list
-        testDialog(Consts.DEFAULT_DECK_ID) { dialog ->
+        // TODO: Update this for new material dialog.
```
see comment above for a fix

david-allison reviewed (changes requested):
Okay, fixed up the unit tests, that one wasn't fun. 

But... pretty much good to go I feel!

ShridharGoel commented on AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.kt:
```diff
@@ -194,14 +194,13 @@ open class CardTemplateEditor : AnkiActivity(), DeckSelectionListener {
-                tempModel = null
-                finishWithAnimation(END)
-            }.build()
+        val discardDialog = DiscardChangesDialog.showDialog(this) {
+            Timber.i("TemplateEditor:: OK button pressed to confirm discard changes")
+            // Clear the edited model from any cache files, and clear it from this objects memory to discard changes
+            TemporaryModel.clearTempModelFiles()
+            tempModel = null
+            finishWithAnimation(END)
+        }
```
`positiveButton` is there in `DiscardChangesDialog.showDialog`.

ShridharGoel commented on AnkiDroid/src/main/java/com/ichi2/anki/noteeditor/Toolbar.kt:
```diff
@@ -220,31 +221,29 @@ class Toolbar : FrameLayout {
-            .items("h1", "h2", "h3", "h4", "h5")
-            .itemsCallback { _: MaterialDialog?, _: View?, _: Int, string: CharSequence ->
-                val formatter = TextWrapper(prefix = "<$string>", suffix = "</$string>")
+        MaterialDialog(context).show {
+            listItems(items = listOf("h1", "h2", "h3", "h4", "h5")) { _: MaterialDialog, _: Int, charSequence: CharSequence ->
+                val formatter = TextWrapper(prefix = "<$charSequence>", suffix = "</$charSequence>")
                 onFormat(formatter)
             }
```
Can you explain this?

ShridharGoel commented on AnkiDroid/src/main/java/com/ichi2/preferences/ControlPreference.kt:
```diff
@@ -83,42 +87,58 @@ class ControlPreference : ListPreference {
-                    onPositive { dialog, _ ->
-                        val gesture = mGesturePicker.getGesture() ?: return@onPositive
+                MaterialDialog(context).show {
+                    title(R.string.binding_add_gesture)
```
Can you explain this?

mikehardy commented:
conflict - this change is huge, so it makes sense but: :weary: - I'll be around a little bit more next day or three, hopefully we can get this in during one of the moments I've got time

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/CardTemplateEditor.kt:
```diff
@@ -741,7 +739,8 @@ open class CardTemplateEditor : AnkiActivity(), DeckSelectionListener {
                 templateFragment
             ) {
-            private var mProgressDialog: MaterialDialog? = null
+            @Suppress("Deprecation")
```
I have not looked but perhaps you know - is this deprecation tracked anywhere?

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:
```diff
@@ -154,7 +152,8 @@ open class DeckPicker :
     private var mBackButtonPressedToExit = false
     private lateinit var mDeckPickerContent: RelativeLayout
-    private var mProgressDialog: MaterialDialog? = null
+    @Suppress("Deprecation") // TODO: Encapsulate ProgressDialog within a class to limit the use of deprecated functionality
```
Okay - now I see more explanation on the deprecation, starts to make more sense, just not sure if it has a tracking issue is the only thing - not a blocker but should be tracked if not

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/ReadText.kt:
```diff
@@ -86,23 +87,23 @@ object ReadText {
-            builder.content(res.getString(R.string.no_tts_available_message))
-                .iconAttr(R.attr.dialogErrorIcon)
-                .positiveText(R.string.dialog_ok)
+            dialog.message(R.string.no_tts_available_message)
```
Log this issue if not logged

david-allison replied:
#11781

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/anki/dialogs/ExportCompleteDialog.kt:
```diff
@@ -39,26 +39,29 @@ class ExportCompleteDialog(private val listener: ExportCompleteDialogListener) :
-            .onNeutral { _: MaterialDialog?, _: DialogAction? -> listener.dismissAllDialogFragments() }
-        return dialogBuilder.show()
+            neutralButton(R.string.dialog_cancel) {
+                // TODO: Discuss regarding alternatives to using a neutral button here
+                //  since it is deprecated and not recommended in material guidelines
```
Please log an issue to track this

david-allison replied:
#11780

mikehardy reviewed (approved):
This looks good enough to go in before it accumulates any more conflicts. My comments are all related to peeling some things off to separate issues, can you please do that so they are tracked in real issues vs just comments in code? Thank you for handling this, I really appreciate it!
